core/backend/src/TxnManager.ts

```diff
@@ -73,12 +77,44 @@ export class TxnManager {
   }
   /** @internal */
   protected _onBeginValidate() { this.validationErrors.length = 0; }
+
+  /** called from native code after validation of a Txn, either from saveChange or apply changeset.

```

Comment:
Is this really called when changesets are applied? If so, do we want to dispatch potentially very large event payloads when pulling changes from hub?

Reply:
yes, it is called for apply changes. We will have to deal with the large set of changes anyway, since an app can do it locally in a single txn too. I was planning on batching the events from native code, somehow. That is, you'd get multiple calls to these events for a single txn. That's why i added the `onEndValidate` event so you can tell when a batch starts.

BTW, we will have to deal with this same problem for geometryChanged events too, right?

Reply:
The recommendation is to end an editing session before pulling changes.

Can listeners distinguish between txns from commit vs apply changes?

Reply:
>The recommendation is to end an editing session before pulling changes.
But what about a txn that makes a lot of local changes?

> Can listeners distinguish between txns from commit vs apply changes?
I'm wrong, we send a different event for OnAppliedChanges. It includes the same lists, so i guess i'll have to make it send a similar event. You need both e.g. to keep selected/highlighted/etc. elements lists valid.

Reply:
added batching, will write tests in next PR